src/Interpreters/ProcessList.cpp

```diff
@@ -373,16 +373,57 @@ QueryStatus::~QueryStatus()
     }
 }
 
+void QueryStatus::ExecutorHolder::cancel()
+{
+    std::lock_guard lock(mutex);
+    if (executor)
+        executor->cancel();
+}
+
+void QueryStatus::ExecutorHolder::remove()
+{
+    std::lock_guard lock(mutex);
+    executor = nullptr;
+}
+
 CancellationCode QueryStatus::cancelQuery(bool)
 {
     if (is_killed.load())
         return CancellationCode::CancelSent;
 
     is_killed.store(true);
 
-    std::lock_guard lock(executors_mutex);
-    for (auto * e : executors)
-        e->cancel();
+    std::unique_lock lock(executors_mutex);
+
+    /// Track all cancelled executors.
+    std::unordered_set<ExecutorHolder *> cancelled;
+    /// We cancel executors from the left to the right, so if the last executor
+    /// was cancelled, then all executors were cancelled.
+    while (!cancelled.contains(executors.back().get()))
+    {
+        size_t size = executors.size();
+        /// We should create a copy of executor holder, because it can be
+        /// removed from vector in removePipelineExecutor from another thread
+        /// and reference will be invalid.
+        for (auto e : executors)
+        {
+            if (cancelled.contains(e.get()))
+                continue;
+            /// We should call cancel() with unlocked executors_mutex, because
+            /// cancel() can try to lock some internal mutex that is already locked by query executing
+            /// thread, and query executing thread can call removePipelineExecutor and lock executors_mutex,
+            /// which will lead to deadlock.
+            lock.unlock();
+            e->cancel();
+            lock.lock();
+            cancelled.insert(e.get());
+            /// While executors_mutex was unlocked, removePipelineExecutor could be called and
+            /// the size of executors could have changed. In this case we should start iterating
+            /// over it again to avoid using invalid iterators.
+            if (executors.size() != size)
```

Comment:
We can move `executors` to a temporary vector, and then iterate it without holding mutex and iterator invalidation.
We need to do it in a loop as well, cause, probably, after releasing a mutex `executors` could be filled with more items.